r/R/dataset.R

```diff
@@ -93,8 +93,19 @@ open_dataset <- function(sources,
     return(dataset___UnionDataset__create(sources, schema))
   }
   factory <- DatasetFactory$create(sources, partitioning = partitioning, ...)
-  # Default is _not_ to inspect/unify schemas
-  factory$Finish(schema, isTRUE(unify_schemas))
+  
+  tryCatch(
+    # Default is _not_ to inspect/unify schemas
+    factory$Finish(schema, isTRUE(unify_schemas)),
+    error = function (e) {
+      msg <- conditionMessage(e)
+      if(grep("Parquet magic bytes not found in footer", msg)){
+        stop("Looks like these are not parquet files, did you mean to specify a 'format'?", call. = FALSE)
```

Comment:
I can see arguments on both sides of this: would it be helpful to include the original error message as well as our new nicer one? 

On the one hand: it's more content and content that some (most?) won't have the context to know what to do with, but on the other it makes clear what the message is form the c++ if someone is familiar with that or does have some reason to suspect that the parquet files are parquet, but were mangled it would tell them that a bit more clearly.

Reply:
Ergh, you make a good point.  I'm tempted to abandon this PR and just solve this via documentation as I don't think there's a good solution in the code.

Reply:
I actually think "just" the message you have here is better than the status quo (and better than docs). But I think you could `paste` together the two messages and be good if you wanted to surface both?

Reply:
I don't like the C++ error messages being exposed to the end user, so I'm with you on that.  I'm now a bit concerned that the error can be triggered by multiple things and it's not easy to work out which - and I don't want the error message to mislead the user.  I'll have a think about if there's a rephrasing that fixes this.

Reply:
Oh, as you mentioned in your initial comment: if the files are Parquet files but mangled.

Reply:
Oh, I think that's totally fine to have a message like this after that. The helpfulness might be a bit of a distraction, but I think most people who get this error message will either be in need of the helpfulness or be knowledgeable enough about their underlying data to not get too distracted by it.

Reply:
So as an exploratory question, "If this message is universally better than the C++ message should the C++ message be changed?  Why or why not?"

Reply:
The error isn't wrong, it's just not informative at this higher layer. "Invalid parquet magic bytes" is meaningful if you yourself called read_parquet(file), but in this case, you weren't intending to read a parquet file because they aren't parquet files.

Reply:
@nealrichardson is 💯 on that, it's also a slightly different audience than that C++ error is targeted at

Reply:
It seems at least there is some room for improvement.  The transition from `read_parquet` to `open_dataset` is at least partially in C++ as well (you are calling `DatasetFactory::Finish` here).  Although the terms are slightly different `e.g. fragments vs sources`.  So I think we could update the C++ error to something like "Dataset creation failed.  The fragment <fragment-to-string> did not match the expected <format> format: <child_error>"

Concretely, "Dataset creation failed.  The fragment '/2019/July/myfile.csv' did not match the expected 'parquet'' format: Parquet magic bytes not found in footer".

I worry this kind of error translation is a slippery slope.

Reply:
I can understand where you're coming from @westonpace - I definitely don't think we should necessarily have error translation for every single C++ error in the package or make that into a pattern we use.

Just to provide a little more context, the reason for opening the original ticket was that when there is a call to `open_dataset()`, if the format is anything other than "parquet", the user needs to pass that format through as an optional argument, but it's not immediately clear from looking at the docs for `open_dataset()` that this is the case.  I'm going to add some examples to the documentation that show how to open files of different formats with `open_dataset()`, but the info in the error is intended as an extra hint in case the behaviour is unintuitive.

Reply:
I've added ARROW-12827 to cleanup the C++ exceptions at some point but I agree that this PR could still be very viable on its own.